File: postgresql/golang/services/postgres_service.go
package services

import (
	"context"
	"database/sql"
	"net/http"
	modelentities "note-golang-postgresql/models/entities"
	modelrequests "note-golang-postgresql/models/requests"
	modelresponses "note-golang-postgresql/models/responses"
	"note-golang-postgresql/repositories"
	"note-golang-postgresql/utils"
)

type PostgresService interface {
	Create(ctx context.Context, createRequest modelrequests.CreateRequest) (response modelresponses.Response)
	Get(ctx context.Context, id int) (response modelresponses.Response)
	Update(ctx context.Context, updateRequest modelrequests.UpdateRequest) (response modelresponses.Response)
	Delete(ctx context.Context, deleteRequest modelrequests.DeleteRequest) (response modelresponses.Response)
}

type postgresService struct {
	PostgresUtil       utils.PostgresUtil
	PostgresRepository repositories.PostgresRepository
}

func NewPostgresService(postgresUtil utils.PostgresUtil, postgresRepository repositories.PostgresRepository) PostgresService {
	return &postgresService{
		PostgresUtil:       postgresUtil,
		PostgresRepository: postgresRepository,
	}
}

func (service *postgresService) Create(ctx context.Context, createRequest modelrequests.CreateRequest) (response modelresponses.Response) {
	tx, err := service.PostgresUtil.BeginTxx(ctx, &sql.TxOptions{})
	if err != nil {
		return modelresponses.SetInternalServerErrorResponse()
	}

	defer func() {
		errCommitOrRollback := service.PostgresUtil.CommitOrRollback(tx, err)
		if errCommitOrRollback != nil {
			response = modelresponses.SetInternalServerErrorResponse()
		}
	}()

	// var user modelentities.User
	// user.Email = createRequest.Email
	// user.Password = createRequest.Password
	var test1 modelentities.Test1
	test1.Test = createRequest.Test
	id, err := service.PostgresRepository.Create(tx, ctx, test1)
	if err != nil {
		return modelresponses.SetInternalServerErrorResponse()
	}
	// user.Id = id
	test1.Id = id
	// modelresponses.SetDataHttpResponse(http.StatusCreated, test1)
	return modelresponses.SetCreateResponse(test1)
}

func (service *postgresService) Get(ctx context.Context, id int) (response modelresponses.Response) {
	test1, err := service.PostgresRepository.Get(service.PostgresUtil.GetDb(), ctx, id)
	if err != nil {
		return modelresponses.SetInternalServerErrorResponse()
	}
	// modelresponses.SetDataHttpResponse(http.StatusCreated, user)
	return modelresponses.SetOkResponse(test1)
}

func (service *postgresService) Update(ctx context.Context, updateRequest modelrequests.UpdateRequest) (response modelresponses.Response) {
	tx, err := service.PostgresUtil.BeginTxx(ctx, &sql.TxOptions{})
	if err != nil {
		return modelresponses.SetInternalServerErrorResponse()
	}

	defer func() {
		errCommitOrRollback := service.PostgresUtil.CommitOrRollback(tx, err)
		if errCommitOrRollback != nil {
			response = modelresponses.SetInternalServerErrorResponse()
		}
	}()

	// var user modelentities.User
	// user.Id = updateRequest.Id
	// user.Email = updateRequest.Email
	// user.Password = updateRequest.Password
	var test1 modelentities.Test1
	test1.Id = updateRequest.Id
	test1.Test = updateRequest.Test
	rowsAffected, err := service.PostgresRepository.Update(tx, ctx, test1)
	if err != nil {
		return modelresponses.SetInternalServerErrorResponse()
	} else if rowsAffected != 1 {
		return modelresponses.SetResponse(http.StatusInternalServerError, nil, modelresponses.MessageResponse{Message: "rows affected not one"})
	}
	// modelresponses.SetDataHttpResponse(http.StatusCreated, user)
	return modelresponses.SetOkResponse(test1)
}

func (service *postgresService) Delete(ctx context.Context, deleteRequst modelrequests.DeleteRequest) (response modelresponses.Response) {
	tx, err := service.PostgresUtil.BeginTxx(ctx, &sql.TxOptions{})
	if err != nil {
		return modelresponses.SetInternalServerErrorResponse()
	}

	defer func() {
		errCommitOrRollback := service.PostgresUtil.CommitOrRollback(tx, err)
		if errCommitOrRollback != nil {
			response = modelresponses.SetInternalServerErrorResponse()
		}
	}()

	rowsAffected, err := service.PostgresRepository.Delete(tx, ctx, deleteRequst.Id)
	if err != nil {
		return modelresponses.SetInternalServerErrorResponse()
	} else if rowsAffected != 1 {
		return modelresponses.SetResponse(http.StatusInternalServerError, nil, modelresponses.MessageResponse{Message: "rows affected create mysql is not 1"})
	}
	// modelresponses.SetDataHttpResponse(http.StatusCreated, modelresponses.SetMessageHttpResponse("successfully delete user"))
	return modelresponses.SetNoContentResponse()
}

File: postgresql/rust/src/services/postgres_service.rs
use std::sync::Arc;

use crate::{models::{entities::test1::Test1, requests::{create_request::CreateRequest, delete_request::DeleteRequest, update_request::UpdateRequest}, responses::{create_response::CreateResponse, get_by_id_response::GetByIdResponse, response::Response, update_response::UpdateResponse}}, repositories::postgres_repository::{PostgresRepository, PostgresRepositoryImpl}, utils::postgres_util::{PostgresUtil, PostgresUtilImpl}};

pub trait PostgresService {
    async fn create(&self, create_request: CreateRequest) -> Response<CreateResponse>;
    async fn get_by_id(&self, id: i32) -> Response<GetByIdResponse>;
    async fn update(&self, update_request: UpdateRequest) -> Response<UpdateResponse>;
    async fn delete(&self, delete_request: DeleteRequest) -> Response<()>;
}

pub struct PostgresServiceImpl {
    postgres_util: Arc<PostgresUtilImpl>,
    postgres_respository: Arc<PostgresRepositoryImpl>
}

impl PostgresServiceImpl {
    pub fn new(postgres_util: Arc<PostgresUtilImpl>, postgres_repository: Arc<PostgresRepositoryImpl>) -> PostgresServiceImpl {
        PostgresServiceImpl {
            postgres_util: postgres_util,
            postgres_respository: postgres_repository
        }
    }
}

impl PostgresService for PostgresServiceImpl {
    async fn create(&self, create_request: CreateRequest) -> Response<CreateResponse> {
        let mut tx = match self.postgres_util.begin().await {
            Ok(tx) => tx,
            Err(err) => {
                println!("err: {}", err);
                return Response::set_internal_server_error_response()
            }
        };

        // if error happened, tx will out of scope and authomatically call rollback
        let mut test1 = Test1 {id: 0, test: create_request.test};
        let id = match self.postgres_respository.create(&mut tx, &test1).await {
            Ok(id) => id,
            Err(err) => {
                println!("err: {}", err);
                // match self.postgres_util.rollback(tx).await {
                //     Ok(()) => (),
                //     Err(err) => {
                //         println!("err: {}", err)
                //         return Response::set_internal_server_error_response();
                //     }
                // }
                return Response::set_internal_server_error_response();
            }
        };
        test1.id = id;

        match self.postgres_util.commit(tx).await {
            Ok(()) => (),
            Err(err) => {
                println!("err: {}", err);
                return Response::set_internal_server_error_response();
            }
        }
        return Response::set_created_response(CreateResponse::set_craeate_response(test1));
    }

    async fn get_by_id(&self, id: i32) -> Response<GetByIdResponse> {
        let test1 = match self.postgres_respository.get_by_id(self.postgres_util.get_pool().await, id).await {
            Ok(test1) => test1,
            Err(err) => {
                println!("err: {}", err);
                return Response::set_internal_server_error_response();
            } 
        };
        return Response::set_ok_response(GetByIdResponse::set_get_by_id_response(test1));
    }

    async fn update(&self, update_request: UpdateRequest) -> Response<UpdateResponse> {
        let mut tx = match self.postgres_util.begin().await {
            Ok(tx) => tx,
            Err(err) => {
                println!("err: {}", err);
                return Response::set_internal_server_error_response();
            }
        };

        let test1 = Test1 {id: update_request.id, test: update_request.test};
        let rows_affected = match self.postgres_respository.update(&mut tx, &test1).await {
            Ok(rows_affected) => rows_affected,
            Err(err) => {
                println!("err: {}", err);
                return Response::set_internal_server_error_response();
            }
        };

        if rows_affected != 1 {
            return Response::set_internal_server_error_response();
        }

        match self.postgres_util.commit(tx).await {
            Ok(()) => (),
            Err(err) => {
                println!("err: {}", err);
                return Response::set_internal_server_error_response();
            }
        }

        return Response::set_ok_response(UpdateResponse::set_update_response(test1))
    }

    async fn delete(&self, delete_request: DeleteRequest) -> Response<()> {
        let mut tx = match self.postgres_util.begin().await {
            Ok(tx) => tx,
            Err(err) => {
                println!("err: {}", err);
                return Response::set_internal_server_error_response();
            }
        };

        let rows_affected = match self.postgres_respository.delete(&mut tx, delete_request.id).await {
            Ok(rows_affected) => rows_affected,
            Err(err) => {
                println!("err: {}", err);
                return Response::set_internal_server_error_response();
            }
        };

        if rows_affected != 1 {
            return Response::set_internal_server_error_response();
        }

        match self.postgres_util.commit(tx).await {
            Ok(()) => (),
            Err(err) => {
                println!("err: {}", err);
                return Response::set_internal_server_error_response();
            }
        }

        return Response::set_no_content_response()
    }
}